src/server/components/dashboard/controllers.ts

```diff
@@ -349,3 +366,167 @@ export async function feedbackController(
+    const annualReviewStartDate = formatAnnualReviewDate((list as List), "annualReviewStartDate");
+    const maxDate = list?.jsonData.annualReviewStartDate ? getMaxDate(list): "";
+    const formattedMaxDate = maxDate ? format(maxDate, DATE_FORMAT) : "";
+
+    res.render("dashboard/lists-edit-annual-review-date", {
+      ...DEFAULT_VIEW_PROPS,
+      error: {
+        text: req.flash('annualReviewError')[0],
+        href: "#annualReviewDateForm",
+      },
+      annualReviewStartDate,
+      maxDate,
+      formattedMaxDate,
+      list,
+      csrfToken: getCSRFToken(req),
+    });
+  } catch(error) {
+    logger.error(`listsEditAnnualReviewDateController Error: ${(error as Error).message}`)
+    next(error);
+  }
+}
+
+function getMaxDate(list: List): Date {
+  const lastAnnualReview = list.jsonData.lastAnnualReviewDate ?? list.createdAt;
+  const lastAnnualReviewPlusYear = add(lastAnnualReview as Date, { years: 1 });
+  const maxDateFromLastAnnualReview = addSixMonths(lastAnnualReviewPlusYear);
+  const maxDateFromUserEnteredValues = addSixMonths(list.jsonData.annualReviewStartDate as number);
+  const maxDate = isBefore(maxDateFromUserEnteredValues, maxDateFromLastAnnualReview)
+    ? maxDateFromUserEnteredValues
+    : maxDateFromLastAnnualReview;
+
+  return maxDate;
+}
+
+function addSixMonths(date: number | string | Date): Date {
+  const annualReviewDate = (typeof date === "string") ? new Date(date) : date;
+  const newDate = add(annualReviewDate as Date, { months: 6 });
+
+  return newDate;
+}
+
+export async function listsEditAnnualReviewDatePostController(
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Promise<void> {
+  try {
+
+    return (req.body.action === "confirmNewDate")
+      ? await confirmNewAnnualReviewDate(req, res)
+      : await updateNewAnnualReviewDate(req, res);
+
+  } catch(error) {
+    logger.error(`listsEditAnnualReviewDatePostController Error: ${(error as Error).message}`)
+    next(error);
+  }
+}
+
+async function confirmNewAnnualReviewDate(req: Request, res: Response): Promise<void> {
+
+  const { listId } = req.params;
+  const list = await findListById(listId) as List;
+  const {day, month} = req.body;
+  const annualReviewDate = getAnnualReviewDate({day, month, list});
+
+  if (!annualReviewDate.isValid) {
+    req.flash('annualReviewError', annualReviewDate.errorMsg);
+    return res.redirect(`${dashboardRoutes.listsEditAnnualReviewDate.replace(":listId", list.id.toString())}`);
+  }
+
+  return res.render("dashboard/lists-edit-annual-review-date-confirm", {
+    ...DEFAULT_VIEW_PROPS,
+    newAnnualReviewDateFormatted: format(annualReviewDate.value as Date, DATE_FORMAT),
+    newAnnualReviewDate: annualReviewDate.value,
+    list,
+    csrfToken: getCSRFToken(req),
+  });
+}
+
+interface IsDateValidInput {
+  day: string,
+  month: string,
+  list: List
+}
+
+interface IsDateValidOutput {
+  isValid: boolean,
+  value: Date | null,
+  errorMsg: string
```

Comment:
I would make errorMsg nullable, if you want to go with this.

But I assume, 
1. if the date is valid, a value is returned?
2. if the date is invalid, there is a message? 

isValid is unnecessary. you code can just check for presence of `isValue`

Reply:
fair point